### ISSUE TITLE
`FLUTTER/bin/internal/shared.sh` presumes NFS is well-behaved for flocks, and breaks otherwise

### ISSUE THREAD

**Issue by RandalSchwartz** [flutter/flutter#79666]
Based on similar issues filed and already closed without resolution, I found that the function `_lock` in `bin/internal/shared.sh` presumes well behaved `flock` functions in Linux.  However, at least Centos with NFS (my setup), the presumption is false, and the function fails.  It works if patched like this:

```
function _lock () {
  # if hash flock 2>/dev/null; then
  #   flock --nonblock --exclusive 7 2>/dev/null
  # elif hash shlock 2>/dev/null; then
  #   shlock -f "$1" -p $$
  # else
  #   mkdir "$1" 2>/dev/null
  # fi
  mkdir "$1" 2>/dev/null
}
```

But that clearly is not a universal patch.

**Comment by jonahwilliams** [comment]
Pull requests welcome!

**Comment by christopherfujino** [comment]
> Based on similar issues filed and already closed without resolution, I found that the function `_lock` in `bin/internal/shared.sh` presumes well behaved `flock` functions in Linux. However, at least Centos with NFS (my setup), the presumption is false, and the function fails. It works if patched like this:
> 
> ```
> function _lock () {
>   # if hash flock 2>/dev/null; then
>   #   flock --nonblock --exclusive 7 2>/dev/null
>   # elif hash shlock 2>/dev/null; then
>   #   shlock -f "$1" -p $$
>   # else
>   #   mkdir "$1" 2>/dev/null
>   # fi
>   mkdir "$1" 2>/dev/null
> }
> ```
> 
> But that clearly is not a universal patch.

I'm guessing the branch your setup is hitting is the first, where `flock` is present but is doing the wrong thing with your NFS?

**Comment by RandalSchwartz** [comment]
> I'm guessing the branch your setup is hitting is the first, where flock is present but is doing the wrong thing with your NFS?

Yes.  It just hangs, usually at the "waiting for other flutter to go away" or something like that, but ps shows no other dart or flutter process.
